Include/stackless_api.h

```diff
@@ -400,32 +400,63 @@ PyAPI_FUNC(int) PyStackless_InitFunctionDeclaration(
+but that would change the function prototypes and thus Python’s C-API. This
+is not practical. Instead, the global variable “_PyStackless_TRY_STACKLESS”
+is used as an implicit parameter.
+
+The content of this variable is moved to the local variable “stackless” at the
+beginning of a C function. In the process, “_PyStackless_TRY_STACKLESS” is set
+to 0, indicating that no unwind-token may be returned. This is done with the
+macro STACKLESS_GETARG() or, for vectorcall functions (see PEP 590), with the
+macro STACKLESS_VECTORCALL_GETARG(), which should be added at the beginning of
+the function declaration.
+
+This design minimizes the possibility of introducing errors due to improper
+return of Py_UnwindToken. The function can contain arbitrary code because the
+flag is hidden in a local variable. If the function is to support soft-
+switching, it must be further adapted. The flag may only be passed to other
+called functions if they adhere to the Stackless-protocol. The macros
+STACKLESS_PROMOTExxx() serve this purpose. To ensure compliance with the
+protocol, the macro STACKLESS_ASSERT() must be called after each such call.
+An exception is the call of vectorcall functions. The call of a vectorcall
+function must be framed with the macros STACKLESS_VECTORCALL_BEFORE() and
+STACKLESS_VECTORCALL_AFTER() or - more simply - performed with the macro
+STACKLESS_VECTORCALL().
+
+Many internal functions have been patched to support this protocol. Their first
+action is a direct or indirect call of the macro STACKLESS_GETARG() or
+STACKLESS_VECTORCALL_GETARG().
+
 
 STACKLESS_GETARG()
 
-  Move the _PyStackless_TRY_STACKLESS flag into the local variable "stackless".
+  Define and initialize the local variable int stackless. The value of
+  stackless is non-zero, if the function may return Py_UnwindToken. After a
+  call to STACKLESS_GETARG() the value of the global variable
+  “_PyStackless_TRY_STACKLESS” is 0.
+
+STACKLESS_VECTORCALL_GETARG(func)
+
+  Vectorcall variant of the macro STACKLESS_GETARG(). Functions of type
+  vectorcallfunc must use STACKLESS_VECTORCALL_GETARG() instead of
+  STACKLESS_GETARG(). The argument "func" must be set to the vectorcall
+  function itself. See function _PyCFunction_FastCallKeywords() for an example.
 
 STACKLESS_PROMOTE_ALL()
 
-  is used for cases where we know that the called function will take
-  care of our object, and we need no test. For example, PyObject_Call
-  and all other Py{Object,Function,CFunction}_*Call* functions use
-  STACKLESS_PROMOTE_xxx, itself, so we don't need to check further.
+  All STACKLESS_PROMOTExxx() macros are used to propagate the stackless-flag
+  from the local variable “stackless” to the global variable
+  “_PyStackless_TRY_STACKLESS”. These macros can’t be used to call a vectorcall
+  function.
+
+  The macro STACKLESS_PROMOTE_ALL() does this unconditionally. It is used for
+  cases where we know that the called function will take care of our object,
```

Comment:
"take care of our object", is confusing.  maybe fix this to say "obeys the stackless protocol by checking the `_PyStackless_TRY_STACKLESS` function and possibly returning the unwind token.  (unless I misunderstand)